fastlane/lanes/build.rb

```diff
@@ -240,169 +240,149 @@
-
-      "#{branch}-#{commit}"
-    end
-  end
-
-  # Posts a comment on the current PR to inform where to download a given Installable Build that was just published to App Center.
-  #
-  # Use this only after `upload_to_app_center` as been called, as it announces how said App Center build can be installed.
-  #
-  # @param [String] app_name The display name to use in the comment text to identify which app this Installable Build is about
-  # @param [String] build_number The App Center's build number for this build
-  # @param [String] url_slug The last component of the `install.appcenter.ms` URL used to install the app. Typically a sluggified version of the app name in App Center (e.g. `WPiOS-One-Offs`)
-  #
-  # @called_by CI — especially, relies on `BUILDKITE_PULL_REQUEST` being defined
-  #
-  def post_installable_build_pr_comment(app_name:, build_number:, url_slug:)
-    UI.message("Successfully built and uploaded installable build `#{build_number}` to App Center.")
-
-    return if ENV['BUILDKITE_PULL_REQUEST'].nil?
-
-    install_url = "https://install.appcenter.ms/orgs/automattic/apps/#{url_slug}/"
-    qr_code_url = "https://chart.googleapis.com/chart?chs=500x500&cht=qr&chl=#{CGI.escape(install_url)}&choe=UTF-8"
-    comment_body = <<~COMMENT_BODY
-      You can test the changes in <strong>#{app_name}</strong> from this Pull Request by:<ul>
-        <li><a href='#{install_url}'>Clicking here</a> or scanning the QR code below to access App Center</li>
-        <li>Then installing the build number <code>#{build_number}</code> on your iPhone</li>
-      </ul>
-      <img src='#{qr_code_url}' width='150' height='150' />
-      If you need access to App Center, please ask a maintainer to add you.
-    COMMENT_BODY
+    # Post PR Comment
+    comment_body = prototype_build_details_comment(
+      app_display_name: output_app_name,
+      app_center_org_name: APPCENTER_OWNER_NAME,
+      metadata: { Configuration: configuration },
+      fold: true
+    )
 
     comment_on_pr(
-      project: 'wordpress-mobile/wordpress-ios',
+      project: GITHUB_REPO,
       pr_number: Integer(ENV.fetch('BUILDKITE_PULL_REQUEST', nil)),
-      reuse_identifier: "installable-build-link--#{url_slug}",
+      reuse_identifier: "prototype-build-link-#{appcenter_app_name}",
       body: comment_body
     )
+
+    # Attach version information as Buildkite metadata and annotation
+    appcenter_id = lane_context.dig(SharedValues::APPCENTER_BUILD_INFORMATION, 'id')
+    metadata = versions.merge(build_type: 'Prototype', 'appcenter:id': appcenter_id)
+    buildkite_metadata(set: metadata)
+    appcenter_install_url = "https://install.appcenter.ms/orgs/#{APPCENTER_OWNER_NAME}/apps/#{appcenter_app_name}/releases/#{appcenter_id}"
+    list = metadata.map { |k, v| " - **#{k}**: #{v}" }.join("\n")
+    buildkite_annotate(context: "appcenter-info-#{output_app_name}", style: 'info', message: "#{output_app_name} [App Center Build](#{appcenter_install_url}) Info:\n\n#{list}")
```

Comment:
Occasionally I noticed there are "Test Failures" annotation in the build page, but all test jobs are successful. I think the root cause is the "Test Failures" annotation isn't removed after a successful retry.

I think a similar issue may occur here too: this "build info" annotation won't be removed if the latest build fails.

Reply:
That fastlane action is called at the very end of the script and buildkite step doing the Prototype build. So an annotation will only be added if the Prototype build step ran all the way to the end once. But if it went all the way to the end without an issue, the CI step will be green and Buildkite won't even show a button to retry the step, so it's impossible for that step to succeed first (and add metadata) then fail later.

(and if the user chooses "Rebuild" on the whole CI build, that creates an entirely separate build and Buildkite page for it anyway)

The only case where I can see the situation that you describes happening is if the script and lane called by that CI step goes all the way to the end, after successfully pushing the build to App Center and adding the annotation, but fail at some meta CI level (eg failure to terminate and unregister the VM properly or whatnot)… and then you retry the step, and then this time if fails before getting to upload a new App Center build and adding the annotation… quite the corner case imho 😅

---

In contrast, the reason why annotations for tests stay around even after build success is a bit different:
 - The test failure annotations are generated on step failure, in case where we can (and are likely to) retry the step. While App Center metadata info annotations are added on success
 - The annotation for test failure annotation is handled by a shell script from our `a8c-ci-toolkit-buildkite-plugin` (`annotate_test_failures` action iirc?) not by the fastlane action from our release-toolkit
 - One of the reasons for why we sometimes still have those annotations reporting test failures even when the step ended up green in the end… might not be because we forgot to remove the annotation, but because the test was flaky but Xcode was set up to do 3 retries before marking it as failed, but still logged every attempt in the `report.junit` (with one XML node for the initial failure and another for the ultimate success for the same test) but our `annotate_test_failures` a8c-ci-toolkit command is not smart enough to pick up on that (we'll have to debug and implement a fix for all that at some point)